### PR TITLE
feat(frontend): plan nested loop temporal join

### PR DESCRIPTION
**Pull request by yuhao-su**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Plan nested loop temporal join for inner and cross join.
- Add e2e and planner test.

The LHS is the stream side. RHS is the look up table. LHS will be broadcasted and RHS is no-shuffle.

We currently disallow left outer join because LHS will be broadcasted and nested loop temporal join will produce duplicated records from all parallelism when it has a NULL match.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

TODO

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/stream_temporal_join.rs [hunk 1]
```diff
@@ -42,12 +42,19 @@ pub struct StreamTemporalJoin {
     core: generic::Join<PlanRef>,
     eq_join_predicate: EqJoinPredicate,
     append_only: bool,
+    is_nested_loop: bool,
```
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/frontend/src/optimizer/plan_node/stream_temporal_join.rs:45]: This is to mark whether StreamTemporalJoin should be executed as nested loop temporal join. Thematically with an equal condition we can still use nested loop here.

File: src/frontend/src/optimizer/plan_node/stream_temporal_join.rs [hunk 2]
```diff
@@ -59,8 +66,14 @@ impl StreamTemporalJoin {
...
-        let l2o = core.l2i_col_mapping().composite(&core.i2o_col_mapping());
-        let dist = l2o.rewrite_provided_distribution(core.left.distribution());
+        let dist = if is_nested_loop {
+            // Use right side distribution directly if it's nested loop temporal join.
+            let r2o = core.r2i_col_mapping().composite(&core.i2o_col_mapping());
+            r2o.rewrite_provided_distribution(core.right.distribution())
+        } else {
+            let l2o = core.l2i_col_mapping().composite(&core.i2o_col_mapping());
+            l2o.rewrite_provided_distribution(core.left.distribution())
+        };
```
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/frontend/src/optimizer/plan_node/stream_temporal_join.rs:76]: But is it going to be the same distribution as we use the left side distribution?

[the remark this answers is not in the document]
Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/stream_temporal_join.rs:76]: A slight difference. The left side distribution is either `Single` or `HashShard`, while the right side distribution could be `UpstreamHashShard`

[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/frontend/src/optimizer/plan_node/stream_temporal_join.rs:76]: In this case, I guess it's better to keep the current plan?

File: src/frontend/src/optimizer/plan_node/stream_temporal_join.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/stream_temporal_join.rs:45]: But I think we should only run a nested loop temporal join when `eq_join_predicate.has_eq() == false`.

File: src/frontend/src/optimizer/plan_node/stream_temporal_join.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/stream_temporal_join.rs:76]: If we keep these two different distribution derivations, it’d be better to document the reason.

File: src/frontend/src/optimizer/plan_node/stream_temporal_join.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/frontend/src/optimizer/plan_node/stream_temporal_join.rs:45]: Okay I'll remove this.

**Review by chenzl25 — APPROVED** [review]
LGTM! thanks!

**Comment by WanYixian** [comment]
Hello @yuhao-su, could you help add release note of it at your earliest convenience, thanks!

**Pull request merged**